textx/model.py

```diff
@@ -598,92 +598,140 @@ def call_obj_processors(metamodel, model_obj,
+                unresolved_count = 1
+                while unresolved_count > 0 and resolved_count > 0:
+                    resolved_count = 0
+                    unresolved_count = 0
+                    # print("***RESOLVING {} models".format(len(models)))
+                    for m in models:
+                        resolved_count_for_this_model, delayed_crossrefs = \
+                            m._tx_reference_resolver.resolve_one_step()
+                        resolved_count += resolved_count_for_this_model
+                        unresolved_count += len(delayed_crossrefs)
+                    # print("DEBUG: delayed #:{} unresolved #:{}".
+                    #      format(unresolved_count,unresolved_count))
+                if (unresolved_count > 0):
+                    error_text = "Unresolvable cross references:"
+
+                    for m in models:
+                        for _, _, delayed \
+                                in m._tx_reference_resolver.delayed_crossrefs:
+                            line, col = parser.pos_to_linecol(delayed.position)
+                            error_text += ' "{}" of class "{}" at {}'.format(
+                                delayed.obj_name, delayed.cls.__name__, (
+                                    line, col))
+                    raise TextXSemanticError(error_text, line=line, col=col)
+
+                for m in models:
+                    # TODO: what does this check?
+                    assert not m._tx_reference_resolver.parser._inst_stack
+
+                # cleanup
+                for m in models:
+                    del m._tx_reference_resolver
+
+                # final check that everything went ok
+                for m in models:
+                    assert 0 == len(get_children_of_type(
+                        Postponed.__class__, m))
+
+                    # We have model loaded and all link resolved
+                    # So we shall do a depth-first call of object
+                    # processors if any processor is defined.
+                    if m._tx_metamodel.obj_processors:
+                        if parser.debug:
+                            parser.dprint("CALLING OBJECT PROCESSORS")
+                        call_obj_processors(m._tx_metamodel, m)
+
+            except BaseException as e:
+                # remove all processed models from (global) repo (if present)
+                # (remove all of them, not only the model with errors,
+                # since, models with errors may be included in other models)
+                for m in models:
+                    for m2 in models:
+                        if hasattr(m2._tx_metamodel, "_tx_model_repository"):
+                            m2._tx_metamodel.\
+                                _tx_model_repository.remove_model(m)
+                        if hasattr(m2, "_tx_model_repository"):
+                            m2._tx_model_repository.remove_model(m)
+                raise e
```

Comment:
This code could be simplified if we introduce `remove_models` which accepts an iterable of models. Then this can be reduced to a simple loop (no nesting). Another issue is that this code is maintaining the cache which is not a concern of `model.py` module but scoping/multi-model support. I suggest making an API in scoping for dealing with this issue and just calling that api from here. That API should also provide a function that would invalidate the cache given  an iterable of changed model paths we discussed in #199.

The same comment holds for the next except block.

Reply:
1. I created **helper functions** to remove models from the cache (and **removed all nested loops**).
2. I simplified the code (I omitted one superfluous deletion).

However, **I did not** introduce the **filename based invalidation** of models. For this I prefer having a look at all aspects of Repositorys in #199 at once and make refactorings (e.g.,  making the map inside the repositories private). 

Here, I **remove models based on the model, not on the file name**. I prefer this here (in principle, I can have "anonymous" entries, see https://github.com/textX/textX/blob/master/textx/scoping/__init__.py#L210, even this should not be possible when including models in our case).

Reply:
Note: I just fixed a bug (only reported for python 3.4 - our luck that python 3.4 has another way to iterate through dicts). --> I reused a filter-object (empty after the first usage) --> solved by converting to a list.

Reply:
@goto40 I agree for file name based invalidation to be part of #199. It should be carefully analyzed and implemented to be easy for usage from textX-LS.

Reply:
Single API call realized by allowing to remove from one single model or a list of models.